core/domain/classifier_domain.py

```diff
@@ -72,11 +116,52 @@ def to_dict(self):
         """
 
         return {
-            'classifier_id': self.id,
-            'exp_id': self.exp_id,
-            'exp_version_when_created': self.exp_version_when_created,
-            'state_name': self.state_name,
-            'algorithm_id': self.algorithm_id,
-            'cached_classifier_data': self.cached_classifier_data,
-            'data_schema_version': self.data_schema_version
+            'classifier_id': self._id,
+            'exp_id': self._exp_id,
+            'exp_version_when_created': self._exp_version_when_created,
+            'state_name': self._state_name,
+            'algorithm_id': self._algorithm_id,
+            'cached_classifier_data': self._cached_classifier_data,
+            'data_schema_version': self._data_schema_version
         }
+
+    def validate(self):
+        """Validates the classifier before it is saved to storage."""
+
+        if not isinstance(self.id, basestring):
+            raise utils.ValidationError(
+                'Expected id to be a string, received %s' % self.id)
+
+        if not isinstance(self.exp_id, basestring):
+            raise utils.ValidationError(
+                'Expected exp_id to be a string, received %s' % self.exp_id)
+        utils.require_valid_name(self.exp_id, 'the exploration id')
+
+        if not isinstance(self.exp_version_when_created, int):
+            raise utils.ValidationError(
+                ('Expected exp_version_when_created to be a integer,') + (
+                    'received %d' % self.exp_version_when_created))
+
+        if not isinstance(self.state_name, basestring):
+            raise utils.ValidationError(
+                'Expected id to be a string, received %s' % self.state_name)
+        utils.require_valid_name(self.state_name, 'the state name')
+
+        if not isinstance(self.algorithm_id, basestring):
+            raise utils.ValidationError(
+                'Expected algorithm_id to be a string, received %s' %(
+                    self.algorithm_id))
+        utils.require_valid_name(
```

Comment:
This seems odd. Should we not just check that the algorithm_id is a valid key for the given interaction id? We'll need some way to get the interaction id though, I think we may need to retrieve the relevant version of the exploration and check its interaction id.